stdlib/public/core/UnsafeRawBufferPointer.swift.gyb

```diff
@@ -348,6 +348,56 @@ public struct Unsafe${Mutable}RawBufferPointer
     return 0
   }
 
+  %  if mutable:
+    /// Initializes memory in the buffer with the elements of
+    /// `source` and binds the initialized memory to type `T`.
+    ///
+    /// Returns an iterator to any elements of `source` that didn't fit in the 
+    /// buffer, and an index to the point in the buffer one past the last element
+    /// written.
+    ///
+    /// - Precondition: The memory in `startIndex..<(source.count *
+    ///   MemoryLayout<T>.stride)` is uninitialized or initialized to a
+    ///   trivial type. The buffer must contain sufficient memory to
+    ///   accommodate at least `source.underestimateCount` elements.
+    ///
+    /// - Postcondition: The memory at `self[startIndex..<returned index] *
+    ///   MemoryLayout<T>.stride` is bound to type `T`.
+    ///
+    /// - Postcondition: The `T` values at `self..<self + source.count *
+    ///   MemoryLayout<T>.stride` are initialized.
+    ///
+    // TODO: Optimize where `C` is a `ContiguousArrayBuffer`.
+    public func initializeMemory<S: Sequence>(
+      as: S.Iterator.Element.Type, from source: S
+    ) -> (S.Iterator, Index) {
+      let underestimate = source.underestimatedCount
+      guard let base = self.baseAddress else {
+        _precondition(underestimate == 0, "no memory available to initialize from source")
+        return (source.makeIterator(),startIndex)
+      }
+      _precondition(underestimate <= count / MemoryLayout<S.Iterator.Element>.stride,
+        "insufficient space to accommodate source.underestimatedCount elements")
+    
+      var it = source.makeIterator()
+      var idx = self.startIndex
+
+      for p in stride(from: base, 
+        // only advance to as far as the last element that will fit
+        to: base + count - (count % MemoryLayout<S.Iterator.Element>.stride), 
+        by: MemoryLayout<S.Iterator.Element>.stride
+      ) {
+        // underflow is permitted – e.g. a sequence into
+        // the spare capacity of an Array buffer
+        guard let x = it.next() else { break }
+        p.initializeMemory(as: S.Iterator.Element.self, to: x)
+        formIndex(&idx, offsetBy: MemoryLayout<S.Iterator.Element>.stride)
+      }
```

Comment:
Optimizing this kind of loop well (so that it lowers to a memcpy when it should) has historically been a really big struggle. Slightly different because it reallocates, but this is the best version the Rust devs have been able to produce, if you're interested in aggressive micro-optimizations: https://doc.rust-lang.org/src/collections/up/src/libcollections/vec.rs.html#1550-1570

If we're serious about it, it might be worth even having an IRGen test verifying that e.g. passing `Array<Int8>` to this lowers to memcpy. (not a blocker, just some idle thoughts)